### PR TITLE
Initial merge of dr-lich into lich5

### PR DESCRIPTION
**Pull request by MahtraDR**

### CODE REVIEW

**Comment by MahtraDR** [comment]
There is some gap I haven't yet identified between the ruby environment my normal lich instances run under, inside a docker container (I'm attachting my Dockerfile for reference) and my host.

Although I am not seeing major differences comparing gem lists, the host is Fedora, the container is Ubuntu. In my container environment, everything just works that I have tested. 

There is some gentle clean up left, but on balance, we're close to being where we just need testers for this.

**Comment by MahtraDR** [comment]
Container setup:
Dockerfile
```
FROM ubuntu:impish

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get -qy install ruby2.7-dev wine libsqlite3-dev \
	build-essential wget git libglib2.0-dev libgirepository1.0-dev \
	libcairo2-dev libpango1.0-dev libgdk-pixbuf2.0-dev libatk1.0-dev \
	libgtk2.0-dev libncurses-dev xterm screen socat iproute2 net-tools

RUN gem install sqlite3 gtk2 logger curses

ADD cmd.sh /tmp

RUN chmod 755 /tmp/cmd.sh

ENTRYPOINT ["/tmp/cmd.sh"]
```
cmd.sh
```
#!/bin/bash
socat TCP-LISTEN:8000,fork,bind=`ip addr show dev tap0 | grep -Po 'inet \K[\d.]+'` TCP:localhost:8000 &

if [ "$1" == '' ]
	then 
		/usr/bin/ruby /home/mahtra/lich/lich.rbw
	else 
		/usr/bin/ruby /home/mahtra/lich/lich.rbw --dragonrealms --login "$1" --without-frontend --detachable-client=8000  --profanity
fi
```

**Comment by MahtraDR** [comment]
Gems inside the container (this works)
```
atk (3.4.3)
benchmark (default: 0.1.0)
bigdecimal (default: 2.0.0)
bundler (default: 2.1.4)
cairo (1.17.5)
cairo-gobject (3.4.3)
cgi (default: 0.1.0)
csv (default: 3.1.2)
curses (1.4.4)
date (default: 3.0.0)
dbm (default: 1.1.0)
delegate (default: 0.1.0)
did_you_mean (default: 1.4.0)
etc (default: 1.1.0)
fcntl (default: 1.0.0)
fiddle (default: 1.0.0)
fileutils (default: 1.4.1)
forwardable (default: 1.3.1)
gdbm (default: 2.1.0)
gdk_pixbuf2 (3.4.3)
getoptlong (default: 0.1.0)
gio2 (3.4.3)
glib2 (3.4.3)
gobject-introspection (3.4.3)
gtk2 (3.4.3)
io-console (default: 0.5.6)
ipaddr (default: 1.2.2)
irb (default: 1.2.6)
json (default: 2.3.0)
logger (1.5.0, default: 1.4.2)
matrix (default: 0.2.0)
minitest (5.13.0)
mutex_m (default: 0.1.0)
native-package-installer (1.1.3)
net-pop (default: 0.1.0)
net-smtp (default: 0.1.0)
net-telnet (0.1.1)
observer (default: 0.1.0)
open3 (default: 0.1.0)
openssl (default: 2.1.2)
ostruct (default: 0.2.0)
pango (3.4.3)
pkg-config (1.4.7)
power_assert (1.1.7)
prime (default: 0.1.1)
pstore (default: 0.1.0)
psych (default: 3.1.0)
racc (default: 1.4.16)
rake (13.0.3)
rdoc (default: 6.2.1.1)
readline (default: 0.0.2)
readline-ext (default: 0.1.0)
red-colors (0.3.0)
reline (default: 0.1.5)
rexml (default: 3.2.3.1)
rss (default: 0.2.8)
rubygems-update (3.2.5)
sdbm (default: 1.0.0)
singleton (default: 0.1.0)
sqlite3 (1.4.2)
stringio (default: 0.1.0)
strscan (default: 1.0.3)
test-unit (3.3.9)
timeout (default: 0.1.0)
tracer (default: 0.1.0)
uri (default: 0.10.0)
webrick (default: 1.6.1)
xmlrpc (0.3.0)
yaml (default: 0.1.0)
zlib (default: 1.1.0)
```

**Comment by MahtraDR** [comment]
gems on host (this does not work)
```
ast (2.4.2)
atk (3.5.1, 3.4.3)
benchmark (0.2.0, default: 0.1.0)
bigdecimal (2.0.0)
bundler (2.1.4)
cairo (1.17.5)
cairo-gobject (3.5.1, 3.4.3)
cgi (default: 0.1.0)
csv (default: 3.1.2)
curses (1.4.4, 1.4.2)
date (default: 3.0.0)
dbm (default: 1.1.0)
delegate (default: 0.1.0)
did_you_mean (default: 1.4.0)
etc (default: 1.1.0)
fcntl (default: 1.0.0)
fiddle (1.1.0, default: 1.0.0)
fileutils (default: 1.4.1)
forwardable (default: 1.3.1)
gdbm (default: 2.1.0)
gdk3 (3.5.1)
gdk_pixbuf2 (3.5.1, 3.4.3)
getoptlong (default: 0.1.0)
gio2 (3.5.1, 3.4.3)
glib2 (3.5.1, 3.4.3)
gobject-introspection (3.5.1, 3.4.3)
gtk2 (3.4.3)
gtk3 (3.5.1)
io-console (0.5.6)
ipaddr (default: 1.2.2)
irb (1.2.6)
json (2.6.1, 2.3.0)
logger (1.5.0, default: 1.4.2)
matrix (0.4.2, default: 0.2.0)
mutex_m (default: 0.1.0)
native-package-installer (1.1.3)
net-pop (default: 0.1.0)
net-smtp (default: 0.1.0)
observer (default: 0.1.0)
open3 (default: 0.1.0)
openssl (2.1.2)
ostruct (0.5.0, default: 0.2.0)
pango (3.5.1, 3.4.3)
parallel (1.21.0)
parser (3.1.0.0)
pkg-config (1.4.7)
prime (default: 0.1.1)
pstore (default: 0.1.0)
psych (3.1.0)
racc (default: 1.4.16)
rainbow (3.0.0)
rake (13.0.1)
rdoc (6.2.1.1)
readline (default: 0.0.2)
readline-ext (default: 0.1.0)
red-colors (0.3.0)
regexp_parser (2.2.0)
reline (default: 0.1.5)
rexml (default: 3.2.3.1)
rss (default: 0.2.8)
rubocop (1.24.1)
rubocop-ast (1.15.1)
ruby-progressbar (1.11.0)
sdbm (default: 1.0.0)
singleton (default: 0.1.0)
sqlite3 (1.4.2)
stringio (default: 0.1.0)
strscan (default: 1.0.3)
thread (0.2.2)
timeout (default: 0.1.0)
tracer (default: 0.1.0)
unicode-display_width (2.1.0)
uri (default: 0.10.0)
webrick (default: 1.6.1)
yaml (default: 0.1.0)
zlib (default: 1.1.0)
```

**Comment by MahtraDR** [comment]
custom_requires works fine now.

The problem is ostruct 0.5.0. Version 0.2.0 everything works, install 0.5.0 and we see the broken bootstrap.

File: lich.rbw [hunk 1]
```diff
@@ -892,9 +900,17 @@ class Script
...
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.lic") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.lic") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.lich") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.lich") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.rb") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.rb") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.cmd") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.cmd") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.wiz") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.wiz") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.lic.gz") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.lic.gz") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.rb.gz") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.rb.gz") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.cmd.gz") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.cmd.gz") ||
+      File.exists?("#{SCRIPT_DIR}/#{script_name}.wiz.gz") || File.exists?("#{SCRIPT_DIR}/custom/#{script_name}.wiz.gz")
```
Comment (rcuhljr) [inline lich.rbw:913]: I feel like `['lic','lich','rb',...].any?{ |ext| a || b }` might be a little saner.

**Review by rpherbig — APPROVED** [review]

**Pull request closed without merging**